Router panics evaluating an AnyOf process posture check when client reports no process/OS data

Affects 2.1.x, 2.0.x, 1.6.x

When an SDK/tunnel client dials a service gated by a process (process-multi) posture check and the client has not reported any process/OS posture data, the router panics with a nil pointer dereference and crashes instead of just denying the dial.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x7ff631353711]

goroutine 2869 [running]:
github.com/openziti/ziti/v2/router/posture.(*ProcessCheck).requireOne(0xab5cea5c8b0, 0xab5ccff0140)
        github.com/openziti/ziti/v2/router/posture/process.go:134 +0x4b1
github.com/openziti/ziti/v2/router/posture.(*ProcessCheck).Evaluate(0xab5cea5c8b0, 0x7ff6327323b8?)
        github.com/openziti/ziti/v2/router/posture/process.go:25 +0x65
github.com/openziti/ziti/v2/router/posture.EvaluatePostureCheck(0x15?, 0xab5ccff0140)
        github.com/openziti/ziti/v2/router/posture/errors.go:71 +0x26
github.com/openziti/ziti/v2/router/posture.IsPassing.func1({0xab5cdf1e6a8, 0x16}, {})
        github.com/openziti/ziti/v2/router/posture/access.go:55 +0x73
github.com/orcaman/concurrent-map/v2.ConcurrentMap[...].IterCb(0x7ff632826d00?, 0xab5ce358eb0?)
        github.com/orcaman/concurrent-map/v2@v2.0.1/concurrent_map.go:293 +0xfa
github.com/openziti/ziti/v2/router/posture.IsPassing(0xab5cea6b290, 0xab5ccff0140)
        github.com/openziti/ziti/v2/router/posture/access.go:47 +0x15a
github.com/openziti/ziti/v2/router/posture.HasAccess(0xab5ce56e3c0, {0xab5ced8e380, 0xa}, {0xab5ceb36048, 0x15}, 0xab5ccff0140, 0x1)
        github.com/openziti/ziti/v2/router/posture/access.go:26 +0x593
github.com/openziti/ziti/v2/router/state.(*ManagerImpl).HasAccess(0xab5cd38ed80?, {0xab5ced8e380, 0xa}, {0xab5ce02ec90?, 0xab5ce3595e8?}, {0xab5ceb36048, 0x15}, 0x1)
        github.com/openziti/ziti/v2/router/state/manager.go:404 +0xa7
github.com/openziti/ziti/v2/router/xgress_edge.(*edgeClientConn).checkAccess(0xab5ceb3a380?, {0xab5ceb36048?, 0xab5cdf5e140?}, 0x1)
        github.com/openziti/ziti/v2/router/xgress_edge/listener.go:1079 +0x8a
github.com/openziti/ziti/v2/router/xgress_edge.(*edgeClientConn).processConnect(0xab5ce854460, 0xab5ce046d80, {0x7ff6327b7520, 0xab5ce29d0e0})
        github.com/openziti/ziti/v2/router/xgress_edge/listener.go:775 +0x3c5
github.com/openziti/ziti/v2/router/xgress_edge.(*Acceptor).BindChannel.func1(0xab5cdf97fd0?, {0x7ff6327b7520?, 0xab5ce29d0e0?})
        github.com/openziti/ziti/v2/router/xgress_edge/accept.go:77 +0x2d
created by github.com/openziti/channel/v4.(*AsyncFunctionReceiveAdapter).HandleReceive in goroutine 3610
        github.com/openziti/channel/v4@v4.3.11/handler.go:112 +0x93
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -120,5 +120,6 @@ are therefore concrete, fixed hostnames, so the set of valid `iss` values stays 
     * [Issue #3684](https://github.com/openziti/ziti/issues/3684) - Keep controller mesh fully connected, as much as possible
     * [Issue #3849](https://github.com/openziti/ziti/issues/3849) - Add a recover mechanism for when a controller cluster can't form a quorum
     * [Issue #3891](https://github.com/openziti/ziti/issues/3891) - Fix OIDC auth failing for controllers using a wildcard server-cert SAN
+    * [Issue #3961](https://github.com/openziti/ziti/issues/3961) - Fix router panic evaluating a process posture check when the client has reported no process/OS posture data
 
 

--- a/router/posture/os.go
+++ b/router/posture/os.go
@@ -22,12 +22,12 @@ func (m *OsCheck) Evaluate(data *InstanceData) *CheckError {
 	}
 
 	osTypeFailure := &OneInListError[Str]{
-		GivenValues: []Str{Str(data.Os.Os.Type)},
+		GivenValues: []Str{Str(data.Os.Os.GetType())},
 	}
 
 	var foundOs *edge_ctrl_pb.DataState_PostureCheck_Os = nil
 	for _, os := range m.OsList {
-		if !strings.EqualFold(strings.ToLower(os.OsType), strings.ToLower(data.Os.Os.Type)) {
+		if !strings.EqualFold(strings.ToLower(os.OsType), strings.ToLower(data.Os.Os.GetType())) {
 			osTypeFailure.ValidValues = append(osTypeFailure.ValidValues, Str(os.OsType))
 		} else {
 			foundOs = os
@@ -43,7 +43,7 @@ func (m *OsCheck) Evaluate(data *InstanceData) *CheckError {
 		}
 	}
 
-	dataVer, err := semver.Make(data.Os.Os.Version)
+	dataVer, err := semver.Make(data.Os.Os.GetVersion())
 
 	if err != nil {
 		return &CheckError{
@@ -54,7 +54,7 @@ func (m *OsCheck) Evaluate(data *InstanceData) *CheckError {
 	}
 
 	osVersionFailure := &OneInListError[Str]{
-		GivenValues: []Str{Str(data.Os.Os.Version)},
+		GivenValues: []Str{Str(data.Os.Os.GetVersion())},
 	}
 
 	for _, version := range foundOs.OsVersions {

--- a/router/posture/process.go
+++ b/router/posture/process.go
@@ -39,8 +39,18 @@ func (p *ProcessCheck) requireAll(data *InstanceData) *CheckError {
 	}
 	cacheProcesses := map[string]*edge_client_pb.PostureResponse_Process{}
 
+	if data.ProcessList == nil {
+		data.ProcessList = &edge_client_pb.PostureResponse_ProcessList{}
+	}
+
 	for _, process := range data.ProcessList.Processes {
 		cacheProcesses[process.Path] = process
+	}
+
+	osType := ""
+
+	if data.Os != nil && data.Os.Os != nil {
+		osType = data.Os.Os.Type
 	}
 
 	allInListError := &AllInListError[*edge_ctrl_pb.DataState_PostureCheck_Process]{}
@@ -64,7 +74,7 @@ func (p *ProcessCheck) requireAll(data *InstanceData) *CheckError {
 			}
 		}
 
-		failedValue := p.compareProcesses(data.Os.Os.Type, cacheProcess, process)
+		failedValue := p.compareProcesses(osType, cacheProcess, process)
 
 		if failedValue != nil {
 			allInListError.FailedValues = append(allInListError.FailedValues, *failedValue)
@@ -74,7 +84,7 @@ func (p *ProcessCheck) requireAll(data *InstanceData) *CheckError {
 	if len(allInListError.FailedValues) > 0 {
 		for _, process := range data.ProcessList.Processes {
 			allInListError.GivenValues = append(allInListError.GivenValues, &edge_ctrl_pb.DataState_PostureCheck_Process{
-				OsType:       data.Os.Os.Type,
+				OsType:       osType,
 				Path:         process.Path,
 				Hashes:       []string{process.Hash},
 				Fingerprints: process.SignerFingerprints,
@@ -131,7 +141,7 @@ func (p *ProcessCheck) requireOne(data *InstanceData) *CheckError {
 
 	for _, process := range data.ProcessList.Processes {
 		anyInList.GivenValues = append(anyInList.GivenValues, &edge_ctrl_pb.DataState_PostureCheck_Process{
-			OsType:       data.Os.Os.Type,
+			OsType:       osType,
 			Path:         process.Path,
 			Hashes:       []string{process.Hash},
 			Fingerprints: process.SignerFingerprints,

--- a/router/posture/process.go
+++ b/router/posture/process.go
@@ -39,18 +39,14 @@ func (p *ProcessCheck) requireAll(data *InstanceData) *CheckError {
 	}
 	cacheProcesses := map[string]*edge_client_pb.PostureResponse_Process{}
 
-	if data.ProcessList == nil {
-		data.ProcessList = &edge_client_pb.PostureResponse_ProcessList{}
-	}
-
-	for _, process := range data.ProcessList.Processes {
+	for _, process := range data.ProcessList.GetProcesses() {
 		cacheProcesses[process.Path] = process
 	}
 
 	osType := ""
 
-	if data.Os != nil && data.Os.Os != nil {
-		osType = data.Os.Os.Type
+	if data.Os != nil {
+		osType = data.Os.Os.GetType()
 	}
 
 	allInListError := &AllInListError[*edge_ctrl_pb.DataState_PostureCheck_Process]{}
@@ -82,7 +78,7 @@ func (p *ProcessCheck) requireAll(data *InstanceData) *CheckError {
 	}
 
 	if len(allInListError.FailedValues) > 0 {
-		for _, process := range data.ProcessList.Processes {
+		for _, process := range data.ProcessList.GetProcesses() {
 			allInListError.GivenValues = append(allInListError.GivenValues, &edge_ctrl_pb.DataState_PostureCheck_Process{
 				OsType:       osType,
 				Path:         process.Path,
@@ -111,11 +107,7 @@ func (p *ProcessCheck) requireOne(data *InstanceData) *CheckError {
 	}
 	cacheProcesses := map[string]*edge_client_pb.PostureResponse_Process{}
 
-	if data.ProcessList == nil {
-		data.ProcessList = &edge_client_pb.PostureResponse_ProcessList{}
-	}
-
-	for _, process := range data.ProcessList.Processes {
+	for _, process := range data.ProcessList.GetProcesses() {
 		cacheProcesses[process.Path] = process
 	}
 
@@ -123,8 +115,8 @@ func (p *ProcessCheck) requireOne(data *InstanceData) *CheckError {
 
 	osType := ""
 
-	if data.Os != nil && data.Os.Os != nil {
-		osType = data.Os.Os.Type
+	if data.Os != nil {
+		osType = data.Os.Os.GetType()
 	}
 
 	for _, process := range p.Processes {
@@ -139,7 +131,7 @@ func (p *ProcessCheck) requireOne(data *InstanceData) *CheckError {
 		}
 	}
 
-	for _, process := range data.ProcessList.Processes {
+	for _, process := range data.ProcessList.GetProcesses() {
 		anyInList.GivenValues = append(anyInList.GivenValues, &edge_ctrl_pb.DataState_PostureCheck_Process{
 			OsType:       osType,
 			Path:         process.Path,

--- a/router/posture/process_test.go
+++ b/router/posture/process_test.go
@@ -28,11 +28,7 @@ func newProcessCheck(semantic string) *ProcessCheck {
 	}
 }
 
-// Regression for the discourse #5881 panic (router crash on dial). The router evaluated a process posture
-// check while the client had reported a running process but NO OS posture data, so InstanceData.Os was nil.
-// The failure-report loop dereferenced data.Os.Os.Type and nil-panicked. The check must instead fail
-// cleanly. The process is reported at the SAME path the check requires, so evaluation reaches the
-// OS-mismatch / report-building code rather than short-circuiting on a missing process.
+// Process reported at the required path but no OS posture: OS state is nil. Must fail cleanly, not panic.
 func TestProcessCheck_ProcessReportedButNilOs_DoesNotPanic(t *testing.T) {
 	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
 		t.Run(semantic, func(t *testing.T) {
@@ -62,8 +58,7 @@ func TestProcessCheck_ProcessReportedButNilOs_DoesNotPanic(t *testing.T) {
 	}
 }
 
-// Empty posture state (neither process list nor OS reported) must also fail cleanly, exercising the
-// nil ProcessList / nil Os guards at the top of the evaluators.
+// No posture data at all: nil ProcessList and nil OS state. Must fail cleanly, not panic.
 func TestProcessCheck_NoPostureData_DoesNotPanic(t *testing.T) {
 	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
 		t.Run(semantic, func(t *testing.T) {
@@ -78,7 +73,7 @@ func TestProcessCheck_NoPostureData_DoesNotPanic(t *testing.T) {
 	}
 }
 
-// A nil InstanceData (no instance at all) must fail cleanly with NilStateError.
+// Nil InstanceData must fail with NilStateError, not panic.
 func TestProcessCheck_NilInstanceData(t *testing.T) {
 	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
 		check := newProcessCheck(semantic)

--- a/router/posture/process_test.go
+++ b/router/posture/process_test.go
@@ -1,0 +1,95 @@
+package posture
+
+import (
+	"testing"
+
+	"github.com/openziti/sdk-golang/pb/edge_client_pb"
+	"github.com/openziti/ziti/v2/common/pb/edge_ctrl_pb"
+	"github.com/openziti/ziti/v2/controller/db"
+)
+
+const testProcPath = "C:\\Windows\\System32\\notepad.exe"
+
+func newProcessCheck(semantic string) *ProcessCheck {
+	return &ProcessCheck{
+		DataState_PostureCheck: &edge_ctrl_pb.DataState_PostureCheck{
+			Id:   "test-id",
+			Name: "my-proc-multi",
+		},
+		DataState_PostureCheck_ProcessMulti: &edge_ctrl_pb.DataState_PostureCheck_ProcessMulti{
+			Semantic: semantic,
+			Processes: []*edge_ctrl_pb.DataState_PostureCheck_Process{
+				{
+					OsType: "Windows",
+					Path:   testProcPath,
+				},
+			},
+		},
+	}
+}
+
+// Regression for the discourse #5881 panic (router crash on dial). The router evaluated a process posture
+// check while the client had reported a running process but NO OS posture data, so InstanceData.Os was nil.
+// The failure-report loop dereferenced data.Os.Os.Type and nil-panicked. The check must instead fail
+// cleanly. The process is reported at the SAME path the check requires, so evaluation reaches the
+// OS-mismatch / report-building code rather than short-circuiting on a missing process.
+func TestProcessCheck_ProcessReportedButNilOs_DoesNotPanic(t *testing.T) {
+	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
+		t.Run(semantic, func(t *testing.T) {
+			check := newProcessCheck(semantic)
+
+			data := &InstanceData{
+				// Os intentionally nil: process posture was sent, OS posture was not.
+				ProcessList: &edge_client_pb.PostureResponse_ProcessList{
+					Processes: []*edge_client_pb.PostureResponse_Process{
+						{
+							Path:      testProcPath,
+							IsRunning: true,
+						},
+					},
+				},
+			}
+
+			result := check.Evaluate(data)
+
+			if result == nil {
+				t.Fatalf("expected a non-nil CheckError when OS posture data is missing, got nil")
+			}
+			if result.Name != "my-proc-multi" {
+				t.Fatalf("expected CheckError for check 'my-proc-multi', got %q", result.Name)
+			}
+		})
+	}
+}
+
+// Empty posture state (neither process list nor OS reported) must also fail cleanly, exercising the
+// nil ProcessList / nil Os guards at the top of the evaluators.
+func TestProcessCheck_NoPostureData_DoesNotPanic(t *testing.T) {
+	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
+		t.Run(semantic, func(t *testing.T) {
+			check := newProcessCheck(semantic)
+
+			result := check.Evaluate(&InstanceData{})
+
+			if result == nil {
+				t.Fatalf("expected a non-nil CheckError when no posture data has been reported, got nil")
+			}
+		})
+	}
+}
+
+// A nil InstanceData (no instance at all) must fail cleanly with NilStateError.
+func TestProcessCheck_NilInstanceData(t *testing.T) {
+	for _, semantic := range []string{db.SemanticAnyOf, db.SemanticAllOf} {
+		check := newProcessCheck(semantic)
+
+		result := check.Evaluate(nil)
+
+		if result == nil {
+			t.Fatalf("semantic %s: expected a non-nil CheckError for nil InstanceData", semantic)
+		}
+		if result.Cause != NilStateError {
+			t.Fatalf("semantic %s: expected NilStateError, got %v", semantic, result.Cause)
+		}
+	}
+}

